src/object/obj_enum.c

```diff
@@ -885,10 +860,176 @@ dss_enum_unpack(vos_iter_type_t type, struct dss_enum_arg *arg,
+	oeda->enum_arg.kds_cap = KDS_NUM;
+	oeda->enum_arg.sgl = &oeda->sgl;
+	oeda->enum_arg.eprs = oeda->eprs;
+	oeda->enum_arg.eprs_cap = KDS_NUM;
+}
+
+void
+daos_enum_dkeys_fini_arg(struct obj_enum_dkeys_arg *oeda)
+{
+	if (oeda->buf != oeda->inline_buf && oeda->buf != NULL)
+		D_FREE(oeda->buf);
+	if (oeda->buf_saved != NULL)
+		D_FREE(oeda->buf_saved);
+}
+
+void
+daos_enum_dkeys_prep_unpack(struct obj_enum_dkeys_arg *oeda)
+{
+	oeda->iov.iov_len = oeda->size;
+	oeda->enum_arg.sgl_idx = 1;
+	oeda->enum_arg.kds_len = oeda->num;
+	oeda->enum_arg.eprs_len = oeda->num;
+}
+
+int
+daos_enum_dkeys_do_list(daos_handle_t oh, daos_epoch_t *epoch,
+			struct obj_enum_dkeys_arg *oeda,
+			daos_obj_list_obj_cb_t list_cb, uint32_t flags)
+{
+	int	rc;
+
+	if (oeda->need_retry) {
+		/* Restore the anchors. */
+		memcpy(&oeda->dkey_anchor, &oeda->dkey_anchor_saved,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->akey_anchor, &oeda->akey_anchor_saved,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->anchor, &oeda->anchor_saved,
+		       sizeof(daos_anchor_t));
+
+		D_ASSERT(oeda->size <= oeda->buf_len);
+
+		oeda->size_saved = oeda->size;
+		if (oeda->size > 0) {
+			if (oeda->buf_len_saved < oeda->size) {
+				if (oeda->buf_saved != NULL)
+					D_FREE(oeda->buf_saved);
+
+				oeda->buf_len_saved = oeda->buf_len;
+				D_ALLOC(oeda->buf_saved, oeda->buf_len_saved);
+				if (oeda->buf_saved == NULL)
+					return -DER_NOMEM;
+			}
+
+			memcpy(oeda->buf_saved, oeda->buf, oeda->size);
+		}
+	} else {
+		/* Backup the anchors. */
+		memcpy(&oeda->dkey_anchor_saved, &oeda->dkey_anchor,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->akey_anchor_saved, &oeda->akey_anchor,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->anchor_saved, &oeda->anchor,
+		       sizeof(daos_anchor_t));
+
+		memset(oeda->kds, 0, sizeof(daos_key_desc_t) * KDS_NUM);
+		memset(oeda->eprs, 0, sizeof(daos_epoch_range_t) * KDS_NUM);
+		memset(oeda->buf, 0, oeda->buf_len);
+
+		oeda->iov.iov_len = 0;
+		oeda->iov.iov_buf = oeda->buf;
+		oeda->iov.iov_buf_len = oeda->buf_len;
+
+		oeda->sgl.sg_nr = 1;
+		oeda->sgl.sg_nr_out = 1;
+
+		oeda->size = 0;
+		oeda->num = KDS_NUM;
+
+		oeda->has_retried = 0;
+		oeda->lost_shard = 0;
+	}
+
+	daos_anchor_set_flags(&oeda->dkey_anchor, flags);
+
+	rc = list_cb(oh, epoch, NULL, NULL, &oeda->size, &oeda->num, oeda->kds,
+		     oeda->eprs, &oeda->sgl, &oeda->anchor, &oeda->dkey_anchor,
+		     &oeda->akey_anchor);
+	if (rc == -DER_KEY2BIG) {
+		D_DEBUG(DB_TRACE, "list obj dkeys on shard "DF_UOID
+			" got -DER_KEY2BIG, key_len "DF_U64"\n",
+			DP_UOID(oeda->enum_arg.oid), oeda->kds[0].kd_key_len);
+		oeda->buf_len = roundup(oeda->kds[0].kd_key_len * 2, 8);
+		if (oeda->buf != oeda->inline_buf)
+			D_FREE(oeda->buf);
+
+		D_ALLOC(oeda->buf, oeda->buf_len);
+		if (oeda->buf == NULL)
+			rc = -DER_NOMEM;
+		else
+			rc = 1;
+	}
+
+	return rc;
+}
+
+int
+daos_enum_dkeys(daos_handle_t oh, daos_unit_oid_t oid, daos_epoch_t epoch,
```

Comment:
hmm, this will enumerate all dkeys/akeys/record, probably should be called enum_obj.   

Btw: if you remove those concurrent io check in verification, you can use this api for object verification as well?

Reply:
We only scan the shard of one object, so enum_obj is more confused, maybe we can rename it as "scan_shard"?

Reply:
> hmm, this will enumerate all dkeys/akeys/record, probably should be called enum_obj.

About the name, I think enum_dkeys may be more suitable, because what we do is the enumeration of dkeys in recursive mode.

Reply:
object iteration is actually iterate everything for the object, IMHO, enum_dkey will cause confusion here.

Reply:
enum_object also can be explained as enumerate the objects in the container?

Reply:
well, it take the object as the input parameter.  and for container iterate, it should be called as container_enum IMHO

Reply:
Hm... I may misunderstand the term of "enumeration". My understand is that "enumerate target" means that we will list all the targets, but NOT list the content of one target.